snapcraft/internal/build_providers/_qemu/_qemu_driver.py

```diff
@@ -0,0 +1,175 @@
+                self._ssh_port)]
+        for drive in qcow2_drives:
+            cmd.append('-drive')
+            cmd.append('file={},if=virtio,format=qcow2'.format(drive))
+        if enable_kvm:
+            cmd.append('-enable-kvm')
+
+        # TODO we might want to spawn another thread here to keep an eye on
+        #      the process. This is good enough for now.
+        self._qemu_proc = _popen(cmd)
+        # Check the immediate return code to make sure things haven't gone
+        # wrong.
+        proc_poll = self._qemu_proc.poll()
+        if proc_poll is not None and proc_poll != 0:
+            raise errors.ProviderLaunchError(
+                provider_name=self.provider_name,
+                exit_code=proc_poll)
+        self._wait_for_ssh()
+
+    def stop(self) -> None:
+        self._ssh_handle.close()
+        try:
+            telnet = telnetlib.Telnet(host='localhost', port=self._telnet_port)
+        except socket.gaierror as telnet_error:
+            raise errors.ProviderCommunicationError(
+                protocol='telnet', port=self._telnet_port,
+                error=telnet_error.strerror) from telnet_error
+        try:
+            telnet.read_until('(qemu) '.encode())
+            telnet.write('savevm latest\n'.encode())
+            telnet.read_until('(qemu) '.encode())
+            telnet.write('quit\n'.encode())
+        except (OSError, EOFError) as telnet_error:
+            raise errors.ProviderStopError(
+                provider_name=self.provider_name,
+                exit_code=None) from telnet_error
+
+    def execute(self, *, command: List[str]) -> None:
+        # Properly quote and join the command
+        command_string = ' '.join([shlex.quote(c) for c in command])
+
+        channel = self._ssh_handle.get_transport().open_session()
+        channel.get_pty()
+        channel.exec_command(command_string)
+        channel.settimeout(0.0)
+
+        exit_code = None
+        while exit_code is None:
+            if channel.recv_ready():
+                sys.stdout.write(channel.recv(1024).decode())
+                sys.stdout.flush()
+            if channel.exit_status_ready():
+                exit_code = channel.recv_exit_status()
+                logger.debug('Command {!r} returned exit code: {}'.format(
+                    command_string, exit_code))
+
+        if exit_code != 0:
+            raise errors.ProviderExecError(
+                provider_name=self.provider_name,
+                command=command,
+                exit_code=exit_code)
+
+    def _wait_for_ssh(self):
+        sock = socket.socket(socket.AF_INET, socket.SOCK_STREAM)
+        ssh_port_listening = False
+        while not ssh_port_listening:
+            time.sleep(1)
+            result = sock.connect_ex(('localhost', self._ssh_port))
+            logger.debug('Pinging for ssh availability: port check {}'.format(
+                result))
+            if result == 0:
+                sock.close()
```

Comment:
Perhaps it would be best to use `with socket.socket(socket.AF_INET, socket.SOCK_STREAM) as socket:` so we don't need to remember to do this.